pkg/ingester/kafka_consumer.go

```diff
@@ -95,35 +97,84 @@ func (kc *kafkaConsumer) consume(ctx context.Context, records []partition.Record
 		minOffset    = int64(math.MaxInt64)
 		maxOffset    = int64(0)
 		consumeStart = time.Now()
+		limitWorkers = kc.maxConsumerWorkers
+		wg           sync.WaitGroup
 	)
 
+	// If the number of workers is set to 0, use the number of available CPUs
+	if limitWorkers == 0 {
+		limitWorkers = runtime.GOMAXPROCS(0)
+	}
+
+	// Find min/max offsets
 	for _, record := range records {
 		minOffset = min(minOffset, record.Offset)
 		maxOffset = max(maxOffset, record.Offset)
 	}
 
 	level.Debug(kc.logger).Log("msg", "consuming records", "min_offset", minOffset, "max_offset", maxOffset)
-	for _, record := range records {
-		stream, err := kc.decoder.DecodeWithoutLabels(record.Content)
-		if err != nil {
-			level.Error(kc.logger).Log("msg", "failed to decode record", "error", err)
-			continue
-		}
-		recordCtx := user.InjectOrgID(record.Ctx, record.TenantID)
-		req := &logproto.PushRequest{
-			Streams: []logproto.Stream{stream},
-		}
-		if err := retryWithBackoff(ctx, func(attempts int) error {
-			if _, err := kc.pusher.Push(recordCtx, req); err != nil {
-				level.Warn(kc.logger).Log("msg", "failed to push records", "err", err, "offset", record.Offset, "attempts", attempts)
-				return err
+
+	numWorkers := min(limitWorkers, len(records))
+	workChan := make(chan partition.Record, numWorkers)
+	// success keeps track of the records that were processed. It is expected to
+	// be sorted in ascending order of offset since the records themselves are
+	// ordered.
+	success := make([]int64, len(records))
+
+	for i := range numWorkers {
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			for record := range workChan {
+				level.Debug(kc.logger).Log("msg", "record", record)
+				stream, err := kc.decoder.DecodeWithoutLabels(record.Content)
+				if err != nil {
+					level.Error(kc.logger).Log("msg", "failed to decode record", "error", err)
+					continue
+				}
+
+				recordCtx := user.InjectOrgID(record.Ctx, record.TenantID)
+				req := &logproto.PushRequest{
+					Streams: []logproto.Stream{stream},
+				}
+
+				level.Debug(kc.logger).Log("msg", "pushing record", "offset", record.Offset, "length", len(record.Content))
+
+				if err := retryWithBackoff(ctx, func(attempts int) error {
+					if _, err := kc.pusher.Push(recordCtx, req); err != nil {
+						level.Warn(kc.logger).Log("msg", "failed to push records", "err", err, "offset", record.Offset, "attempts", attempts)
+						return err
+					}
+					return nil
+				}); err != nil {
+					level.Error(kc.logger).Log("msg", "exhausted all retry attempts, failed to push records", "err", err, "offset", record.Offset)
+					continue
+				}
+
+				success[i] = record.Offset
```

Comment:
You are overwriting the offsets of other records here. In the code example in the issue, `i` corresponds to the index of the record in `records`, not the index of the worker. You can fix this by passing the index to the goroutine via `workChan`, or passing the offsets back in a results channel.

Reply:
thanks for checking, changed now to send the record and index together